src/System.Management.Automation/engine/NativeCommandParameterBinder.cs

```diff
@@ -207,105 +227,103 @@ private void appendOneNativeArgument(ExecutionContext context, object obj, char
-                                    }
-                                }
-                                else
-                                {
-                                    _arguments.Append(arg);
-                                }
+                    // Expand paths, but only from the file system.
+                    if (paths?.Count > 0 && paths.All(p => p.BaseObject is FileSystemInfo))
+                    {
+                        var sep = "";
+                        foreach (var path in paths)
+                        {
+                            _arguments.Append(sep);
+                            sep = " ";
+                            var expandedPath = (path.BaseObject as FileSystemInfo).FullName;
+                            if (normalizePath)
+                            {
+                                expandedPath =
+                                    Context.SessionState.Path.NormalizeRelativePath(expandedPath, cwdinfo.ProviderPath);
+                            }
+                            // If the path contains spaces, then add quotes around it.
+                            if (NeedQuotes(expandedPath))
+                            {
+                                _arguments.Append("\"");
+                                _arguments.Append(expandedPath);
+                                _arguments.Append("\"");
                             }
                             else
                             {
-                                // Even if there are no wildcards, we still need to possibly
-                                // expand ~ into the filesystem provider home directory path
-                                ProviderInfo fileSystemProvider = Context.EngineSessionState.GetSingleProvider(
-                                    Microsoft.PowerShell.Commands.FileSystemProvider.ProviderName);
-                                string home = fileSystemProvider.Home;
-                                if (string.Equals(arg, "~"))
-                                {
-                                    _arguments.Append(home);
-                                }
-                                else if (arg.StartsWith("~/", StringComparison.OrdinalIgnoreCase))
-                                {
-                                    var replacementString = home + arg.Substring(1);
-                                    _arguments.Append(replacementString);
-                                }
-                                else
-                                {
-                                    _arguments.Append(arg);
-                                }
+                                _arguments.Append(expandedPath);
                             }
-#else
-                            _arguments.Append(arg);
-#endif
+                            argExpanded = true;
                         }
                     }
                 }
-            } while (list != null);
+            }
+            else
```

Comment:
Now we will expand `~` even if `arg` is not a bare word. Do we want to expand `~` unconditionally?

Reply:
Another question: it seems to me that after this change we don't glob for variables. Is it intentional?
For example, before the change
```
PS> $var = '*.txt'
PS> /bin/echo $var                                                                               
a.txt b.txt c.txt
```
with the change
```
PS> /bin/echo $var                                                                                
*.txt
```

Reply:
Thanks for catching, I missed this when reviewing Bruce's changes, but you're right, we don't want to expand `~` unconditionally. I'll fix it.

Reply:
And I see bash also expands variables unless quoted, so we should probably match that behavior too.

Reply:
@daxian-dbw - updated. If you sign off, I'll merge - I want to squash the last 2 commits and rebase 2 commits instead squashing the 3 commits.